src/pykeen/nn/node_piece/anchor_search.py

```diff
@@ -222,6 +224,175 @@ def __call__(self, edge_index: numpy.ndarray, anchors: numpy.ndarray, k: int) ->
+        yield from super().extra_repr()
+        yield f"max_iter={self.max_iter}"
+
+    @staticmethod
+    def create_adjacency(
+        edge_index: numpy.ndarray,
+    ) -> torch.tensor:
+        """
+        Create a sparse adjacency matrix (in the form of the edge list) from a given edge index.
+
+        :param edge_index: shape: (2, m)
+            the edge index
+
+        :return: shape: (2, 2m + n)
+            edge list with inverse edges and self-loops
+        """
+        # infer shape
+        num_entities = edge_index.max().item() + 1
+        edge_index = torch.as_tensor(edge_index, dtype=torch.long)
+
+        # symmetric + self-loops
+        edge_list = torch.cat(
+            [edge_index, edge_index.flip(0), torch.arange(num_entities).unsqueeze(0).repeat(2, 1)], dim=-1
+        ).unique(
+            dim=1
+        )  # unique for deduplicating repeated edges
+
+        return edge_list
+
+    @staticmethod
+    def bfs(
+        anchors: numpy.ndarray,
+        edge_list: torch.tensor,
+        max_iter: int,
+        k: int,
+        device: torch.device,
+    ) -> numpy.ndarray:
+        """
+        Determine the candidate pool using breadth-first search.
+
+        :param anchors: shape: (a,)
+            the anchor node IDs
+        :param edge_list: shape: (2, n)
+            the edge list with symmetric edges and self-loops
+        :param max_iter:
+            the maximum number of hops to consider
+        :param k:
+            the minimum number of anchor nodes to reach
+        :param device:
+            the device on which the calculations are done
+
+        :return: shape: (n, a)
+            a boolean array indicating whether anchor $j$ is in the set of $k$ closest anchors for node $i$
+
+        :raises ImportError:
+            If :mod:`torch_sparse` is not installed
+        """
+        try:
+            import torch_sparse
+        except ImportError as err:
+            raise ImportError("Requires `torch_sparse` to be installed.") from err
+
+        num_entities = edge_list.max().item() + 1
+        # for each entity, determine anchor pool by BFS
+        num_anchors = len(anchors)
+
+        anchors = torch.tensor(anchors, dtype=torch.long, device=device)
+
+        # an array storing whether node i is reachable by anchor j
+        reachable = torch.zeros((num_entities, num_anchors), dtype=torch.bool, device=device)
+        reachable[anchors] = torch.eye(num_anchors, dtype=torch.bool, device=device)
+
+        # an array indicating whether a node is closed, i.e., has found at least $k$ anchors
+        final = torch.zeros((num_entities,), dtype=torch.bool, device=device)
+
+        # the output that track the distance to each found anchor
+        # dtype is unsigned int 8 bit, so we initialize the maximum distance to 255 (or max default)
+        # TODO int8 takes 8x more space than just a binary matrix, decide if we want to keep it
+        dtype = torch.uint8
```

Comment:
do we want to parametrize this?

Reply:
rather not, the main `pool` matrix is supposed to take as small space as possible (after the `bool` type, so it's `uint8`), and for all sane cases (eg, not running a BFS for 128+ iterations) this will cover all our needs

Reply:
okay so we decided to keep it, i guess we can delete line 320